Superdense Coding Kata: Add EntangledPair task (#1408)
microsoft/qsharp: katas/content/superdense_coding/index.md

# Superdense Coding Kata

@[section]({
    "id": "./superdense_coding__overview",
    "title": "Overview"
})

Superdense coding protocol allows us to transmit two bits of classical information by sending just one qubit using previously shared quantum entanglement. We are asssuming that Alice is the sender and Bob is the receiver.

- A good description can be found in [the Wikipedia article](https://en.wikipedia.org/wiki/Superdense_coding).
- A great interactive demonstration can be found [on the Wolfram Demonstrations Project](https://demonstrations.wolfram.com/SuperdenseCoding/).

We split the superdense coding protocol into several steps:

- Preparation (creating the entangled pair of qubits that are sent to Alice and Bob).
- Encoding the message (Alice's task): Encoding the classical bits of the message into the state of Alice's qubit which then is sent to Bob.
- Decoding the message (Bob's task): Using Bob's original qubit and the qubit he received from Alice to decode the classical message sent.
- Finally, we compose those steps into the complete superdense coding protocol.

@[exercise]({
    "id": "./superdense_coding__entangled_pair",

>     "title": "Entangled pair",

Reviewer: ```suggestion
    "title": "Entangled Pair",
```

> # Superdense Coding Kata

Reviewer: When you create a new kata, you need to add it to index.json - otherwise CI build will not build it. (Then, if it's still a work in progress, you remove the addition from the list before merging the PR, so that it doesn't get published accidentally.)

Author: Done. I tested it locally before pushing :-)

I was wondering if its worth having a WIP list of katas that get tested in CI and local but not get published. That will give us immunity from accidental publishing :-)

@[section]({

>     "id": "./superdense_coding.ipynb__overview",

Reviewer: ```suggestion
    "id": "./superdense_coding__overview",
```

    "title": "Overview"
})

> **Superdense Coding** quantum kata is a series of exercises designed to get you familiar with programming in Q#.
> 
> It covers the superdense coding protocol which allows us to transmit two bits of classical information by sending just one qubit using previously shared quantum entanglement.

Reviewer: We tend to drop the "this is a series of exercises" part of the intro in the new tutorials, and jump right into the topic, especially for a kata that is that late in the series - if the learner made it so far, they probably got the idea :-)

```suggestion
Superdense coding protocol allows us to transmit two bits of classical information by sending just one qubit using previously shared quantum entanglement.
```

> - A good description can be found in [the Wikipedia article](https://en.wikipedia.org/wikiSuperdense_coding).

Reviewer: ```suggestion
- A good description can be found in [the Wikipedia article](https://en.wikipedia.org/wiki/Superdense_coding).
```

> - A great interactive demonstration can be found [on the Wolfram Demonstrations Project](http:/demonstrations.wolfram.com/SuperdenseCoding/).

Reviewer: ```suggestion
- A great interactive demonstration can be found [on the Wolfram Demonstrations Project](https://demonstrations.wolfram.com/SuperdenseCoding/).
```

> - Superdense coding protocol is described in Nielsen & Chuang, section 2.3 (pp. 97-98).

Reviewer: I would drop this one - it's not a link, and we cannot assume the reader bought a book too.
```suggestion
```

Author: Done, in earlier Katas, we were referring this book a lot. I like this book, it really explains some of the concepts in great details.

> We split the superdense coding protocol into several steps, following the description in the [Wikipedia article](https://en.wikipedia.org/wiki/Superdense_coding):

Reviewer: ```suggestion
We split the superdense coding protocol into several steps:
```

> - Preparation (creating the entangled pair of qubits that are sent to Alice and Bob).

Reviewer: Using Alice and Bob here is common, but it also implies that the learner already read the other sources. I would rename them to sender and receiver to make it more descriptive.

Author: Thanks for the suggestion. We were following this notation in previous kata. I added a note about Alice being the sender and bob being the receiver. Let me know if that works, or should I go more descriptive Alice(sender), Bob(receiver)?

Personally speaking, I like Alice(typical sender), Bob(typical receiver) and Charlie/Eve(Evesdropper). This is because I find it to be more cognitive distinguishable vs sender/receiver/evesdropper.
